keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,2169 @@
+    // +optional
+    PodSelector *metav1.LabelSelector
+    // PodPriority specifies a pod priority.
+    // Pods with a priority less or equal to this value are selected.
+    PodPriority int32
+    // PodType selects pods according to the pod type:
+    // - Default selects all pods except DaemonSet and Static pods.	
+    // - DaemonSet selects DaemonSet pods.
+    // - Static selects static pods.
+    PodType PodType
+}
+
+type NodeMaintenanceStatus struct {
+    // StageStatuses tracks the statuses of started stages.
+    StageStatuses []StageStatus
+    // List of a maintenance statuses for all nodes targeted by this maintenance.
+    NodeStatuses []NodeMaintenanceNodeStatus
+    Conditions []metav1.Condition
+}
+
+type StageStatus struct {
+    // Name of the Stage.
+    Name NodeMaintenanceStage
+    // StartTimestamp is the time that indicates the start of this stage.
+    StartTimestamp *metav1.Time
+}
+
+type NodeReference struct {
+    // Name of the node.
+    Name string
+}
+
+type NodeMaintenanceNodeStatus struct {
+    // NodeRef identifies a Node.
+    NodeRef NodeReference
+    // DrainTargets specifies which pods on this node are currently being targeted for evacuation.
+    // Once evacuation of the Default PodType finishes, DaemonSet PodType entries appear.
+    // Once the evacuation of DaemonSet PodType finishes, Static PodType entries appear.
+    // The PodPriority for these entries is increased over time according to the .spec.DrainPlan
+    // as the lower-priority pods finish evacuation.
+    // The next entry in the .spec.DrainPlan is selected once all the nodes have reached their
+    // DrainTargets.
+    // If there are multiple NodeMaintenances for a node, the least powerful DrainTargets among
+    // them are selected and set for all the NodeMaintenances for that node. Thus, the DrainTargets
+    // do not have to correspond to the entries in .spec.drainPlan for a single NodeMaintenance
+    // instance.
+    // DrainTargets cannot backtrack and will target more pods with each update until all pods on
+    // the node are targeted.
+    DrainTargets  []DrainPlanEntry
+    // DrainMessage may specify a state of the drain on this node and a reason why the drain
+    // targets are set to a particular values.
+    DrainMessage string
+    // Number of pods that have not yet started evacuation.
+    PodsPendingEvacuation int32
+    // Number of pods that have started evacuation and have a matching Evacuation object.
+    PodsEvacuating int32
+}
+
+const (
+    // DrainedCondition is a condition set by the node-maintenance controller that signals
+    // whether all pods pending termination have terminated on all target nodes when drain is
+    // requested by the maintenance object.
+    DrainedCondition = "Drained"
+}
+```
+
+### NodeMaintenance Admission
+
+`nodemaintenance` admission plugin will be introduced.
+
+It will validate all incoming requests for CREATE, UPDATE, and DELETE operations. All nodes
```

Comment:
on the on the `NodeMaintainance` object, right?

Reply:
yes, updated